pkg/natdiscovery/remote_endpoint_test.go

```diff
@@ -0,0 +1,164 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natdiscovery
+
+import (
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("natdiscovery listener", func() {
+
+	var rnat *remoteEndpointNAT
+
+	BeforeEach(func() {
+		remoteEndpoint := createTestRemoteEndpoint()
+		rnat = newRemoteEndpointNAT(&remoteEndpoint)
+	})
+
+	When("created", func() {
+		It("should report as shouldCheck", func() {
+			Expect(rnat.shouldCheck()).To(BeTrue())
+		})
+	})
+
+	When("created, and right after have checked", func() {
+		It("should not report as shouldCheck", func() {
+			rnat.checkSent()
+			Expect(rnat.shouldCheck()).To(BeFalse())
+		})
+	})
+
+	When("more than totalTimeout seconds passed", func() {
+		It("should report as timed out", func() {
+			rnat.started = time.Now().Add(-totalTimeout)
+			Expect(rnat.hasTimedOut()).To(BeTrue())
+		})
+	})
+
+	When("legacy mode is triggered", func() {
+		It("should select the private IP for NATEnabled is false", func() {
+			rnat.Endpoint.Spec.NATEnabled = false
+			rnat.useLegacyNATSettings()
+			Expect(rnat.state).To(Equal(selectedPrivateIP))
+			Expect(rnat.UseIP).To(Equal(rnat.Endpoint.Spec.PrivateIP))
+		})
+
+		It("should select the public IP for NATEnabled is true", func() {
+			rnat.Endpoint.Spec.NATEnabled = true
+			rnat.useLegacyNATSettings()
+			Expect(rnat.state).To(Equal(selectedPublicIP))
+			Expect(rnat.UseIP).To(Equal(rnat.Endpoint.Spec.PublicIP))
+		})
+	})
+
+	When("an public IP is selected, but no check was sent", func() {
+		It("it should fail", func() {
+			err := rnat.transitionToPublicIP(testRemoteEndpointName, false)
+			Expect(err).To(HaveOccurred())
+			Expect(rnat.UseIP).To(Equal(""))
+		})
+	})
+
+	When("an private IP is selected, but no check was sent", func() {
+		It("it should fail", func() {
+			err := rnat.transitionToPrivateIP(testRemoteEndpointName, false)
+			Expect(err).To(HaveOccurred())
+			Expect(rnat.UseIP).To(Equal(""))
+		})
+	})
+
+	When("public IP is selected", func() {
+		It("should use the public IP", func() {
+			rnat.checkSent()
+			err := rnat.transitionToPublicIP(testRemoteEndpointName, false)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(rnat.state).To(Equal(selectedPublicIP))
+			Expect(rnat.UseIP).To(Equal(rnat.Endpoint.Spec.PublicIP))
+			Expect(rnat.UseNAT).To(BeFalse())
+		})
+	})
+
+	When("private IP is selected", func() {
+		It("should use the private IP", func() {
+			rnat.checkSent()
+			err := rnat.transitionToPrivateIP(testRemoteEndpointName, false)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(rnat.state).To(Equal(selectedPrivateIP))
+			Expect(rnat.UseIP).To(Equal(rnat.Endpoint.Spec.PrivateIP))
+			Expect(rnat.UseNAT).To(BeFalse())
+		})
+	})
```

Comment:
The tests on L134 below are similar and can be combined:

```suggestion
	When("the private IP is selected", func() {
		var useNAT bool

		JustBeforeEach(func() {
			rnat.checkSent()
			err := rnat.transitionToPrivateIP(testRemoteEndpointName, useNAT)
			Expect(err).NotTo(HaveOccurred())
			Expect(rnat.state).To(Equal(selectedPrivateIP))
		})

		It("should use the private IP", func() {
			Expect(rnat.UseIP).To(Equal(rnat.Endpoint.Spec.PrivateIP))
		})

		Context("with NAT enabled", func() {
			BeforeEach(func() {
				useNAT = true
			})

			It("should set useNAT to true", func() {
				Expect(rnat.UseNAT).To(BeTrue())
			})
		})

		Context("with NAT disabled", func() {
			BeforeEach(func() {
				useNAT = false
			})

			It("should set useNAT to false", func() {
				Expect(rnat.UseNAT).To(BeFalse())
			})
		})
	})
```